embrace-android-delivery/src/main/kotlin/io/embrace/android/embracesdk/internal/delivery/execution/RequestExecutionServiceImpl.kt

```diff
@@ -1,15 +1,62 @@
 package io.embrace.android.embracesdk.internal.delivery.execution
 
+import android.util.Log
 import io.embrace.android.embracesdk.internal.comms.api.ApiResponse
+import io.embrace.android.embracesdk.internal.comms.api.Endpoint
 import io.embrace.android.embracesdk.internal.delivery.SupportedEnvelopeType
+import okhttp3.Headers.Companion.toHeaders
+import okhttp3.OkHttpClient
+import okhttp3.Protocol
+import okhttp3.Request
+import okhttp3.RequestBody.Companion.toRequestBody
 import java.io.InputStream
 
-class RequestExecutionServiceImpl : RequestExecutionService {
+class RequestExecutionServiceImpl(
+    private val coreBaseUrl: String,
+    private val lazyDeviceId: Lazy<String>,
+    private val appId: String,
+    private val embraceVersionName: String,
+) : RequestExecutionService {
+
+    private val okHttpClient = OkHttpClient()
+        .newBuilder()
+        .protocols(listOf(Protocol.H2_PRIOR_KNOWLEDGE))
+        .build()
 
     override fun attemptHttpRequest(
         payloadStream: () -> InputStream,
-        envelopeType: SupportedEnvelopeType
+        envelopeType: SupportedEnvelopeType,
     ): ApiResponse {
-        return ApiResponse.None
+        val apiRequest = envelopeType.endpoint.getApiRequestFromEndpoint()
+
+        val payload: ByteArray = payloadStream().use {
```

Comment:
Not a massive fan of loading the entire payload into a `ByteArray` as it means memory usage linearly increases with the size of the payload. Does `OkHttp` provide a mechanism that allows a stream (or Okio) equivalent to be used instead? That will have a fixed buffer so memory usage won't grow excessively.

Reply:
Yeah, it's what's worrying me the most, I'll try to find some alternative

Reply:
There's a recipe here that might be useful as a starting point, it'll probably be simplest to write to the `OutputStream` as suggested: https://square.github.io/okhttp/recipes/#post-streaming-kt-java